File: 0784_Letter Case Permutation/LetterCasePermutation.java
class Solution {
    public List<String> letterCasePermutation(String S) {
        List<String> ans = new ArrayList<String>();
        DFS(S.toCharArray(), 0, ans);
        return ans;
    }
    
    private void DFS(char[] c_arr, int s,  List<String> ans){
        if(s == c_arr.length){
            ans.add(new String(c_arr));
            return;
        }
        
        DFS(c_arr, s+1, ans);
        if(!Character.isLetter(c_arr[s])) return;
        c_arr[s] ^= 1 << 5;
        DFS(c_arr, s+1, ans);
        c_arr[s] ^= 1 << 5;
    }
}

File: 0784_Letter Case Permutation/LetterCasePermutation.py
class Solution(object):
    def letterCasePermutation(self, S):
        """
        :type S: str
        :rtype: List[str]
        """
        ans = []
        
        def dfs(S,i,n):
            if i == n:
                ans.append("".join(S))
                return
            dfs(S,i+1,n)
            if not S[i].isalpha(): return
            S[i] = chr(ord(S[i]) ^ (1<<5))
            dfs(S,i+1,n)
            S[i] = chr(ord(S[i]) ^ (1<<5))
        
        dfs(list(S),0,len(S))
        return ans